### PR TITLE
feat(route): add cors header in route.fulfill

### PR DESCRIPTION
**Pull request by yury-s**
#12929

### CODE REVIEW

File: packages/playwright-core/src/server/network.ts [hunk 1]
```diff
@@ -265,9 +265,18 @@ export class Route extends SdkObject {
         isBase64 = false;
       }
     }
+    const headers = [...(overrides.headers || [])];
+    if (overrides.cors !== false) {
+      const corsHeader = headers.find(({ name, value }) => name === 'access-control-allow-origin');
+      // See https://github.com/microsoft/playwright/issues/12929
+      if (!corsHeader) {
+        const origin = this._request.headerValue('origin') || '*';
+        headers.push({ name: 'access-control-allow-origin', value: origin });
```
Comment (dgozman) [inline packages/playwright-core/src/server/network.ts:274]: MDN suggests to also send `Vary: Origin`: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#access-control-allow-origin.

File: packages/playwright-core/src/server/network.ts [hunk 2]
```diff
@@ -265,9 +265,18 @@ export class Route extends SdkObject {
         isBase64 = false;
       }
     }
+    const headers = [...(overrides.headers || [])];
+    if (overrides.cors !== false) {
```
Comment (dgozman) [inline packages/playwright-core/src/server/network.ts:269]: I guess we never arrive here for preflights? If we do, we might need some special handling.

File: packages/playwright-core/src/server/network.ts [hunk 3]
```diff
@@ -265,9 +265,18 @@ export class Route extends SdkObject {
...
+      const corsHeader = headers.find(({ name, value }) => name === 'access-control-allow-origin');
+      // See https://github.com/microsoft/playwright/issues/12929
+      if (!corsHeader) {
+        const origin = this._request.headerValue('origin') || '*';
+        headers.push({ name: 'access-control-allow-origin', value: origin });
+      }
```
Comment (dgozman) [inline packages/playwright-core/src/server/network.ts:275]: Should we also set `Access-Control-Allow-Credentials: true`?

File: docs/src/api/class-route.md [hunk 4]
```diff
@@ -226,6 +226,13 @@ is resolved relative to the current working directory.
 
 [APIResponse] to fulfill route's request with. Individual fields of the response (such as headers) can be overridden using fulfill options.
 
+### option: Route.fulfill.cors
+- `cors` <[boolean]>
```
Comment (dgozman) [inline docs/src/api/class-route.md:230]: WDYT about `cors: 'allow' | 'deny' | 'none'`?

Also, here is an example of configuring cors on the server side (similar to our use case): https://expressjs.com/en/resources/middleware/cors.html

File: packages/playwright-core/src/server/network.ts (hunk 2 above)
Comment (yury-s) [reply packages/playwright-core/src/server/network.ts:269]: Added a test, we don't currently intercept preflights in any browser.

File: packages/playwright-core/src/server/network.ts [hunk 5]
```diff
@@ -264,9 +265,24 @@ export class Route extends SdkObject {
...
+      const corsHeader = headers.find(({ name, value }) => name === 'access-control-allow-origin');
+      // See https://github.com/microsoft/playwright/issues/12929
+      if (!corsHeader) {
+        const origin = this._request.headerValue('origin');
+        if (origin) {
+          headers.push({ name: 'access-control-allow-origin', value: origin });
+          headers.push({ name: 'access-control-allow-credentials', value: 'true' });
+          headers.push({ name: 'vary', value: 'Origin' });
+        } else {
+          headers.push({ name: 'access-control-allow-origin', value: '*' });
```
Comment (dgozman) [inline packages/playwright-core/src/server/network.ts:279]: Let's drop this one.

**Review by dgozman — APPROVED** [review]

**Pull request merged**